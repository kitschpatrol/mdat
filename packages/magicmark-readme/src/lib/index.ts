export {
	type CheckReadmeFileOptions,
	type CheckReadmeFileReport,
	type CheckReadmeStringOptions,
	type CheckReadmeStringReport,
	checkReadmeFile,
	checkReadmeString,
} from './check'

export {
	type ExpandReadmeFileOptions,
	type ExpandReadmeFileReport,
	type ExpandReadmeStringOptions,
	type ExpandReadmeStringReport,
	expandReadmeFile,
	expandReadmeString,
} from './expand'
