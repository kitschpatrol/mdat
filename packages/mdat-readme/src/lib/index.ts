export {
	type CleanReadmeConfig,
	type CleanReadmeFileReport,
	type ExpandReadmeConfig,
	type ExpandReadmeFileReport,
	type ExpandReadmeRules,
	type ExpandReadmeStringReport,
	cleanReadmeFile,
	cleanReadmeString,
	expandReadmeFile,
	expandReadmeString,
	getPackageJson,
	getReadmeConfig,
} from './api'
export { type MdatReadmeConfig, loadConfigReadme } from './config'
