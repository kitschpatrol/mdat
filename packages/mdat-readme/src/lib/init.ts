// This will do something soon
console.log('Hello, world!')
