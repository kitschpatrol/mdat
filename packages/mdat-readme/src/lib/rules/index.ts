import badges from './badges'
import banner from './banner'
import cliHelp from './cli-help'
import code from './code'
import contributing from './contributing'
import footer from './footer'
import header from './header'
import license from './license'
import shortDescription from './short-description'
import tableOfContents from './table-of-contents'
import title from './title'
import toc from './toc'
import { type Rules } from 'remark-mdat'

// A little redundant to have an expander name / preset in both
// the record keys and the `keyword` field in the expansion object,
// but this smooths over the kebab vs. camel case difference and arguably
// makes expansions more portable

export default {
	...badges,
	...banner,
	...cliHelp,
	...code,
	...contributing,
	...footer,
	...header,
	...license,
	...shortDescription,
	...tableOfContents,
	...toc,
	...title,
} satisfies Rules
