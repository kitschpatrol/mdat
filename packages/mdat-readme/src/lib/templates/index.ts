// Bundle template markdown for convenience

// Note:
// This project is build with tsup, not vite... but we've configured a custom
// loader for .md files in tsup.config.ts and a module declaration in
// ./src/types/md.d.ts. We use the ?raw suffix to intersect with the vite loader
// and get the raw file contents in the vitest context, without actually having
// to include vite as a direct dev dependency.

import readmeBasicCompound from './readme-basic-compound.md?raw'
import readmeBasic from './readme-basic.md?raw'
import readmeFullCompound from './readme-full-compound.md?raw'
import readmeFull from './readme-full.md?raw'

export default {
	readmeBasic,
	readmeBasicCompound,
	readmeFull,
	readmeFullCompound,
}
