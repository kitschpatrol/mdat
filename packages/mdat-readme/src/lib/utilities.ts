import { findUp } from 'find-up'
import { log } from 'mdat'
import fs from 'node:fs/promises'
import path from 'node:path'
import { packageUp } from 'package-up'
import { packageDirectory } from 'pkg-dir'
import { type NormalizedPackageJson, readPackage } from 'read-pkg'

/**
 * Searches for a readme file in the following order:
 * 1. Looks for readme.md in the closest package directory
 * 2. Searches up from the current working directly for readme.md
 *
 * @returns The path to the readme file
 * @throws If no readme is found
 */
export async function findReadme(): Promise<string> {
	log.info(`Searching for package directory...`)

	// First, check for a readme in the package directory
	const thePackageDirectory = await packageDirectory()
	if (thePackageDirectory !== undefined) {
		log.info(`Found package directory: "${thePackageDirectory}"`)
		const files = await fs.readdir(thePackageDirectory)
		const readmeFile = files.find((file) => file.toLowerCase() === 'readme.md')
		if (readmeFile !== undefined) {
			return path.join(thePackageDirectory, readmeFile)
		}

		log.warn(`Readme not found in package directory, searching up for readme...`)
	}

	// FindUp is case-insensitive
	const closestReadme = await findUp('readme.md', {
		type: 'file',
	})

	if (closestReadme !== undefined) {
		log.info(`Found closest readme at "${closestReadme}"`)
		return closestReadme
	}

	throw new Error('No readme found')
}

// Decadent
// function getCasePermutations(input: string, index = 0, current = ''): string[] {
// 	if (index === input.length) {
// 		return [current]
// 	}

// 	const char = input[index]
// 	const nextPermutations = getCasePermutations(input, index + 1, current + char)

// 	if (char.toLowerCase() !== char.toUpperCase()) {
// 		const upperPermutations = getCasePermutations(input, index + 1, current + char.toUpperCase())
// 		return [...nextPermutations, ...upperPermutations]
// 	}

// 	return nextPermutations
// }

export async function findPackage(): Promise<string> {
	const packageFile = await packageUp()
	if (packageFile === undefined) {
		throw new Error('No package.json found')
	}

	return packageFile
}

// Careful, global stuff here...
// For use by rules, allows override of package.json
// via cli options
let packageJson: NormalizedPackageJson | undefined
let packageFile: string | undefined

export function setPackageFile(path: string) {
	packageFile = path
}

// Load as package json only as needed, memoize
export async function getPackageJson(): Promise<NormalizedPackageJson> {
	// Find package as needed
	packageFile ??= await findPackage()
	packageJson ??= await readPackage({ cwd: path.dirname(packageFile) })

	if (packageJson === undefined) {
		throw new Error('No package.json found')
	}

	return packageJson
}

// Sometimes rules need the path to the package.json
export async function getPackagePath(): Promise<string> {
	// Find package as needed
	packageFile ??= await findPackage()
	return packageFile
}
