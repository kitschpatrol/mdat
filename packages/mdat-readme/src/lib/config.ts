import readmeRules from './rules'
import { findPackage, findReadme } from './utilities'
import { type MdatConfig, loadConfig } from 'mdat'
import { z } from 'zod'

export type MdatReadmeConfig = {
	assetsPath?: string
	packageFile?: string
	readmeFile?: string
} & MdatConfig

const mdatReadmeConfigExtensionSchema = z
	.object({
		assetsPath: z.string().optional(),
		packageFile: z.string().optional(),
		readmeFile: z.string().optional(),
	})
	.describe('Mdat Readme Config Extension')

async function getReadmeDefaults(): Promise<MdatReadmeConfig> {
	return {
		addMetaComment: true,
		assetsPath: './assets',
		packageFile: await findPackage(),
		readmeFile: await findReadme(),
		rules: readmeRules,
	}
}

// Convenience loader to always include the default readme config
// Note passing config type to loadConfig parameter plucker, and zod schema to
// represent the additional keys in mdat-readme specific configuration files
type LoadConfigOptions = Parameters<typeof loadConfig<MdatReadmeConfig>>[0]
export async function loadConfigReadme(options?: LoadConfigOptions): Promise<MdatReadmeConfig> {
	const { additionalConfig = [], configExtensionSchema, ...rest } = options ?? {}
	const readmeDefaultConfig = await getReadmeDefaults()
	const additionalConfigArray = Array.isArray(additionalConfig)
		? additionalConfig
		: [additionalConfig]

	const result = await loadConfig<MdatReadmeConfig>({
		additionalConfig: [readmeDefaultConfig, ...additionalConfigArray],
		configExtensionSchema:
			configExtensionSchema === undefined
				? mdatReadmeConfigExtensionSchema
				: mdatReadmeConfigExtensionSchema.merge(configExtensionSchema),
		...rest,
	})

	return result
}
