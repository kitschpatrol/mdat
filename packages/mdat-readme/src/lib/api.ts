import { type MdatReadmeConfig, loadConfigReadme } from './config'
import { setPackageFile } from './utilities'
import { type ExpandConfig, type ExpandRules, expandFile, expandString } from 'mdat'
import type { Simplify } from 'type-fest'
import { type VFile } from 'vfile'

export type ExpandReadmeConfig = ExpandConfig<MdatReadmeConfig>
export type ExpandReadmeRules = ExpandRules // No change

// Expand a readme file, maps to CLI
export type ExpandReadmeFileReport = Simplify<
	ExpandReadmeStringReport & {
		readmeFile: string
	}
>

export async function expandReadmeFile(
	config?: ExpandReadmeConfig,
	rules?: ExpandReadmeRules,
): Promise<ExpandReadmeFileReport> {
	const resolvedConfig = await loadConfigReadme({
		additionalConfig: config,
		additionalRules: rules,
	})
	const { packageFile, readmeFile, ...expandFileConfig } = resolvedConfig

	// This should never happen because the defaults are set in loadConfigReadme
	if (packageFile === undefined || readmeFile === undefined) {
		throw new Error('Package and readme files are required')
	}

	// Hacky global state for rules
	setPackageFile(packageFile)

	// No second deep load because expandFileOptions already has loaded and normalized rules
	const result = await expandFile(readmeFile, undefined, undefined, expandFileConfig)

	return {
		packageFile,
		readmeFile,
		result,
	}
}

// Expand a readme string, mostly for testing
export type ExpandReadmeStringReport = {
	packageFile: string
	result: VFile
}

export async function expandReadmeString(
	markdown: string,
	config?: ExpandReadmeConfig,
	rules?: ExpandReadmeRules,
): Promise<ExpandReadmeStringReport> {
	const resolvedConfig = await loadConfigReadme({
		additionalConfig: config,
		additionalRules: rules,
	})
	const { packageFile, readmeFile, ...expandConfig } = resolvedConfig

	// This should never happen because the defaults are set
	if (packageFile === undefined || readmeFile === undefined) {
		throw new Error('Package and readme files are required')
	}

	// Hacky global state for rules
	setPackageFile(packageFile)

	const result = await expandString(markdown, expandConfig)

	return {
		packageFile,
		result,
	}
}
