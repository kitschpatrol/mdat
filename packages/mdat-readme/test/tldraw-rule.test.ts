import { expandReadmeString } from '../src/lib/api'
import os from 'node:os'
import { describe, expect, it } from 'vitest'

// These have long timeouts because going through puppeteer is slow

describe('tldraw image rule', () => {
	it('should expand tldraw images from local files', async () => {
		const markdown = `<!-- tldraw { src: "./test/assets/tldraw-sketch.tldr" } -->`
		const result = await expandReadmeString(markdown, {
			assetsPath: `${os.tmpdir()}/assets`,
		})
		expect(result).toMatchSnapshot()
	}, 30_000)

	it('should expand tldraw images from remote URLs', async () => {
		const markdown = `<!-- tldraw { src: "https://www.tldraw.com/s/v2_c_JsxJk8dag6QsrqExukis4" } -->`
		const result = await expandReadmeString(markdown, {
			assetsPath: `${os.tmpdir()}/assets`,
		})
		expect(result).toMatchSnapshot()
	}, 30_000)
})
