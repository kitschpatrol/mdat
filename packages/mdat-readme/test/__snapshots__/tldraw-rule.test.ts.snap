// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tldraw image rule > should expand tldraw images from local files 1`] = `
{
  "packageFile": "/Users/mika/Code/mdat/packages/mdat-readme/package.json",
  "result": VFile {
    "cwd": "/Users/mika/Code/mdat/packages/mdat-readme",
    "data": {},
    "history": [],
    "messages": [
      [1:1-1:1: Expanded: <!-- tldraw { src: "./test/assets/tldraw-sketch.tldr" } -->],
      [1:1-1:1: Missing required: <!-- contributing -->],
      [1:1-1:1: Missing required: <!-- license -->],
      [1:1-1:1: Missing required: <!-- short-description -->],
      [1:1-1:1: Missing required: <!-- title -->],
    ],
    "value": "<!--+ Warning: Content in HTML comment blocks generated by mdat on 2024-02-07 +-->

<!-- tldraw { src: "./test/assets/tldraw-sketch.tldr" } -->

<picture>
  <source media="(prefers-color-scheme: dark)" srcset="/var/folders/7w/vbgxhpks5pz_9z06ms_4s9f80000gn/T/assets/tldraw-sketch-132cbdb8-dark.svg">
  <source media="(prefers-color-scheme: light)" srcset="/var/folders/7w/vbgxhpks5pz_9z06ms_4s9f80000gn/T/assets/tldraw-sketch-132cbdb8-light.svg">
  <img alt="tldraw diagram" src="/var/folders/7w/vbgxhpks5pz_9z06ms_4s9f80000gn/T/assets/tldraw-sketch-132cbdb8-light.svg">
</picture>

<!-- /tldraw -->
",
  },
}
`;

exports[`tldraw image rule > should expand tldraw images from remote URLs 1`] = `
{
  "packageFile": "/Users/mika/Code/mdat/packages/mdat-readme/package.json",
  "result": VFile {
    "cwd": "/Users/mika/Code/mdat/packages/mdat-readme",
    "data": {},
    "history": [],
    "messages": [
      [1:1-1:1: Expanded: <!-- tldraw { src: "https://www.tldraw.com/s/v2_c_JsxJk8dag6QsrqExukis4" } -->],
      [1:1-1:1: Missing required: <!-- contributing -->],
      [1:1-1:1: Missing required: <!-- license -->],
      [1:1-1:1: Missing required: <!-- short-description -->],
      [1:1-1:1: Missing required: <!-- title -->],
    ],
    "value": "<!--+ Warning: Content in HTML comment blocks generated by mdat on 2024-02-07 +-->

<!-- tldraw { src: "https://www.tldraw.com/s/v2_c_JsxJk8dag6QsrqExukis4" } -->

<picture>
  <source media="(prefers-color-scheme: dark)" srcset="/var/folders/7w/vbgxhpks5pz_9z06ms_4s9f80000gn/T/assets/v2_c_JsxJk8dag6QsrqExukis4-c1827e44-dark.svg">
  <source media="(prefers-color-scheme: light)" srcset="/var/folders/7w/vbgxhpks5pz_9z06ms_4s9f80000gn/T/assets/v2_c_JsxJk8dag6QsrqExukis4-c1827e44-light.svg">
  <img alt="tldraw diagram" src="/var/folders/7w/vbgxhpks5pz_9z06ms_4s9f80000gn/T/assets/v2_c_JsxJk8dag6QsrqExukis4-c1827e44-light.svg">
</picture>

<!-- /tldraw -->
",
  },
}
`;
